src/java/htsjdk/samtools/util/BlockCompressedInputStream.java

```diff
@@ -515,6 +590,172 @@ private static boolean preambleEqual(final byte[] preamble, final byte[] buf, fi
+         */
+        private volatile BlockingQueue<DecompressedBlock> mResult = null;
+        private AtomicBoolean mTerminate = new AtomicBoolean(false);
+        
+        /**
+         * Calling thread entry point indicating that the background
+         * decompression thread should stop
+         */
+        public void terminate() {
+            mTerminate.set(true);
+            mOperations.clear();
+        }
+        
+        /**
+         * Calling thread entry point indicating that no more eager
+         * background decompression tasks should be run.
+         */
+        public void ensureStopped() {
+            mResult = null;
+            mOperations.clear();
+        }
+        
+        /**
+         * Calling thread entry point indicating that background
+         * decompression should be happening at the current stream position
+         */
+        public void ensureStarted() {
+            if (mResult == null) {
+                mResult = new ArrayBlockingQueue<DecompressedBlock>(READ_AHEAD_BUFFERS);
+                assert(mOperations.isEmpty());
+                for (int i = 0; i < READ_AHEAD_BUFFERS; i++) {
+                    queueDecompressionTask(null);
+                }
+            }
+        }
+        
+        /**
+         * Calling thread entry point indicating that more blocks
+         * should be decompressed
+         * @param buffer decompression buffer to be reused if possible
+         */
+        public void queueDecompressionTask(byte[] buffer) {
+            if (buffer == null) {
+                // ArrayBlockingQueue does not allow null
+                mOperations.add(NO_ASYNC_REUSE_BUFFER_AVAILABLE);
+            } else {
+                mOperations.add(buffer);
+            }
+        }
+        
+        /**
+         * Calling thread entry point to retrieve the next decompressed block. 
+         * @return next decompressed block
+         * @throws InterruptedException
+         */
+        public DecompressedBlock takeNextDecompressedBlock() throws InterruptedException {
+            return mResult.take();
+        }
+        /**
+         * Background thread worker loop
+         */
+        private void decompressBlocks() {
+            try {
+                while (true) {
+                    byte[] buffer = mOperations.take();
+                    // snapshot where to output to. If seek() is called
+                    // on the input stream, then under no circumstances
+                    // do we want to write our result to the post-seek
+                    // results
+                    BlockingQueue<DecompressedBlock> result = mResult;
+                    if (result != null) {
+                        DecompressedBlock decompressed = nextBlock(buffer);
+                        result.put(decompressed);
+                    }
+                }
+            } catch (InterruptedException e) {
+                // Reset interrupt status
+                Thread.interrupted();
```

Comment:
why is this necessary - wouldn't the interrupted flag be reset by whoever threw the exception? Is it necessary because the main thread may be calling `interrupt()` on the worker after the `InterruptedException` is thrown but before it caught (?)